services/sources/publisher_backlist_service.py

```diff
@@ -0,0 +1,115 @@
+
+class PublisherBacklistService(SourceService):
+    def __init__(self):
+        self.airtable_auth_token = os.environ.get('AIRTABLE_KEY', None)
+
+    def get_records(
+        self,
+        full_import: bool=False, 
+        start_timestamp: datetime=None,
+        offset: Optional[int]=None,
+        limit: Optional[int]=None
+    ) -> list[UofMMapping]:
+        
+        array_json_records = self.get_records_json(full_import, start_timestamp, offset, limit)
+
+        array_index = 0
+        while array_index != len(array_json_records):
+            for dict_index in range(0, len(array_json_records[array_index]['records'])):
+                try:
+                    record_metadata_dict = array_json_records[array_index]['records'][dict_index]
+                    record = UofMMapping(record_metadata_dict)
+                    record.applyMapping()
+                except Exception as e:
+                    logger.exception(f'Failed to process Publisher Backlist record due to {e}')
+
+            array_index += 1
+        
+
+
+    def get_records_json(self,
+        full_import: bool=False, 
+        start_timestamp: datetime=None,
+        offset: Optional[int]=None,
+        limit: Optional[int]=None
+    ) -> list[dict]:
+
+        if offset == None:
+            limit = 100
+
+        headers = {"Authorization": f"Bearer {self.airtable_auth_token}"}
+
+        if not full_import:
+            if start_timestamp:
+
+                filter_by_formula = self.build_filter_by_formula_parameter(start_timestamp)
+
+                array_json_records = self.get_records_array(headers, limit, filter_by_formula)
+
+                return array_json_records
+
+            else:
+
+                filter_by_formula = self.build_filter_by_formula_parameter(start_timestamp)
+
+                array_json_records = self.get_records_array(headers, limit, filter_by_formula)
+
+                return array_json_records
+            
+        else:
+
+            array_json_records = self.get_records_array(headers, limit, filter_by_formula=None)
+        
+            return array_json_records
+        
+    def build_filter_by_formula_parameter(self, start_timestamp: datetime=None) -> str:
+        if not start_timestamp:
+            start_timestamp = datetime.now(timezone.utc).replace(tzinfo=None) - timedelta(hours=24)
+
+        start_date_time_str = start_timestamp.strftime("%Y-%m-%d %H:%M:%S.%f")
+        start_date_time_encoded = urllib.parse.quote(start_date_time_str)
+        is_same_date_time_filter = f"IS_SAME(%7BLast%20Modified%7D,%20%22{start_date_time_encoded}%22,%20%22second%22"
+        is_after_date_time_filter = f"%20IS_AFTER(%7BLast%20Modified%7D,%20%22{start_date_time_encoded}%22"
+        
+        filter_by_formula = f"OR({is_same_date_time_filter}),{is_after_date_time_filter}))"
+        return filter_by_formula
```

Comment:
Nice - this looks good. Do the Airtable last modified dates have seconds to them as well? Could we simplify this to do is after without the seconds or even minutes? 

Nit: recommend adding a newline between the body of this function and the return statement for readability.

Reply:
The Airtable last modified dates do have seconds in their response like this: `2024-06-20T17:22:51.000Z` I just tested out the IS_SAME filter in Insomnia and unit of time like seconds aren't required in the filter parameter which isn't stated in the Airtable formulas sheet on their website unlike other filters that detail when a parameter is optional. The unit of time details how we should match the dates up to a certain unit like hours, minutes, and seconds. Therefore, I can get rid of the unit of time parameter in that filter to simplify the filter.

Reply:
nice!